src/scan-imports.ts

```diff
@@ -1,159 +1,131 @@
-      return (node.node.arguments[0] as any).value;
-    }
+  // check if the resolved specifier (including file extension) is a known package.
+  const resolvedSpecifier = specifier.substring(webModulesIndex + 'web_modules/'.length);
+  if (knownDependencies.includes(resolvedSpecifier)) {
+    return resolvedSpecifier;
   }
-  return undefined;
-}
-
-/**
- * Return array of dependencies
- */
-function parseImports(file: string) {
-  const code = fs.readFileSync(file, 'utf8');
-
-  // skip file if empty, or if permissions error
-  if (!code) {
-    return [];
+  // check if the resolved specifier (without extension) is a known package.
+  const resolvedSpecifierWithoutExtension = stripJsExtension(resolvedSpecifier);
+  if (knownDependencies.includes(resolvedSpecifierWithoutExtension)) {
+    return resolvedSpecifierWithoutExtension;
   }
-  const deps = new Set<string>();
+  // Otherwise, this is an explicit import to a file within a package.
+  return resolvedSpecifier;
+}
 
-  // try to parse; this will fail if not JS
+function getInstallTargetsForFile(code: string, knownDependencies: string[]): InstallTarget[] {
+  const allImports: InstallTarget[] = [];
   try {
     const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
     traverse(ast, {
-      enter(node) {
-        const moduleName =
-          importModuleName(node as NodePath<ImportDeclaration>) ||
-          dynamicImportModuleName(node as NodePath<CallExpression>);
-
-        if (moduleName) {
-          deps.add(moduleName);
+      ImportDeclaration(path) {
+        const webModuleSpecifier = parseWebModuleSpecifier(
+          path.node.source.value,
+          knownDependencies,
+        );
+        if (webModuleSpecifier) {
+          allImports.push({
+            specifier: webModuleSpecifier,
+            all: false,
+            default: path.node.specifiers.some(s => s.type === 'ImportDefaultSpecifier'),
+            namespace: path.node.specifiers.some(s => s.type === 'ImportNamespaceSpecifier'),
+            named: path.node.specifiers
+              .map(s => s.type === 'ImportSpecifier' && s.imported.name)
+              .filter(Boolean),
+          });
+        }
+      },
+      Import(path) {
+        // Only match dynamic imports that are called as a function
+        if (path.parent.type !== 'CallExpression') {
+          return;
+        }
+        // Only match dynamic imports called with a single string argument
+        const [argNode] = path.parent.arguments;
+        if (argNode.type !== 'StringLiteral') {
+          return;
+        }
+        // Analyze that string argument as an import specifier
+        const webModuleSpecifier = parseWebModuleSpecifier(argNode.value, knownDependencies);
+        if (webModuleSpecifier) {
+          allImports.push(createInstallTarget(webModuleSpecifier, true));
         }
       },
     });
-    return [...deps];
   } catch (e) {
+    // TODO: report the error
```

Comment:
What will fail here is the `parse()` function if it encounters invalid JavaScript, or the user included non-JS files in their glob. I don’t think anything else would fail in this block, so if this does report something it might be just a simple “Skipping [file]” warning to the user since it can’t be parsed (read: probably isn’t JS).

Reply:
+1 Whoops, I forgot to come back to this. Will add that "skipping" message.